### PR TITLE
resolver-requester for full history nodes

### PR DESCRIPTION
**Pull request by mihaib79**

### CODE REVIEW

File: dataRetriever/mock/currentNetworkEpochProviderMock.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package mock
+
+import (
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+)
+
+type CurrentNetworkEpochProviderStub struct {
```
Comment (iulianpascalau) [inline dataRetriever/mock/currentNetworkEpochProviderMock.go:9]: This is actually a mock as it have functionality that resembles the "real" component. Also, missing comment like `CurrentNetworkEpochProviderStub -`

File: dataRetriever/interface.go [hunk 2]
```diff
@@ -359,3 +361,10 @@ type ResolverDebugHandler interface {
 	LogSucceededToResolveData(topic string, hash []byte)
 	IsInterfaceNil() bool
 }
+
+type CurrentNetworkEpochProviderHandler interface {
```
Comment (iulianpascalau) [inline dataRetriever/interface.go:365]: Missing comment

File: p2p/libp2p/netMessenger.go [hunk 3]
```diff
@@ -637,6 +637,20 @@ func (netMes *networkMessenger) ConnectedPeersOnTopic(topic string) []core.PeerI
 	return netMes.poc.ConnectedPeersOnChannel(topic)
 }
 
+// ConnectedFullHistoryPeersOnTopic returns the connected peers on a provided topic
+func (netMes *networkMessenger) ConnectedFullHistoryPeersOnTopic(topic string) []core.PeerID {
```
Comment (iulianpascalau) [inline p2p/libp2p/netMessenger.go:641]: missing unit test

File: dataRetriever/resolvers/topicResolverSender/topicResolverSender.go [hunk 4]
```diff
@@ -105,20 +118,29 @@ func (trs *topicResolverSender) SendOnRequestTopic(rd *dataRetriever.RequestData
...
 
-	crossPeers := trs.peerListCreator.PeerList()
-	numSentCross := trs.sendOnTopic(crossPeers, topicToSendRequest, buff, trs.numCrossShardPeers)
+	var numSentIntra, numSentCross int
+	var intraPeers, crossPeers []core.PeerID
+	fullHistoryPeers := make([]core.PeerID, 0)
 
-	intraPeers := trs.peerListCreator.IntraShardPeerList()
-	numSentIntra := trs.sendOnTopic(intraPeers, topicToSendRequest, buff, trs.numIntraShardPeers)
+	if trs.currentNetworkEpochProviderHandler.EpochIsActiveInNetwork(rd.Epoch) {
```
Comment (iulianpascalau) [inline dataRetriever/resolvers/topicResolverSender/topicResolverSender.go:125]: missing unit tests

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: dataRetriever/errors.go [hunk 5]
```diff
@@ -155,6 +155,9 @@ var ErrBadRequest = errors.New("request should not be done as it doesn't follow
 // ErrNilAntifloodHandler signals that a nil antiflood handler has been provided
 var ErrNilAntifloodHandler = errors.New("nil antiflood handler")
 
+// ErrNilCurrentNetworkEpochProvider signals that a nil antiflood handler has been provided
```
Comment (AdoAdoAdo) [inline dataRetriever/errors.go:158]: error description incorrect

File: dataRetriever/mock/currentNetworkEpochProviderMock.go (hunk 1 above)
Comment (AdoAdoAdo) [reply dataRetriever/mock/currentNetworkEpochProviderMock.go:9]: +1 please change the struct name to CurrentNetworkEpochProviderMock

File: dataRetriever/resolvers/topicResolverSender/diffPeerListCreator.go [hunk 6]
```diff
@@ -73,6 +73,10 @@ func (dplc *DiffPeerListCreator) IntraShardPeerList() []core.PeerID {
 	return dplc.messenger.ConnectedPeersOnTopic(dplc.intraShardTopic)
 }
 
+func (dplc *DiffPeerListCreator) FullHistoryList() []core.PeerID {
```
Comment (AdoAdoAdo) [inline dataRetriever/resolvers/topicResolverSender/diffPeerListCreator.go:76]: please add comments to exported methods.

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: dataRetriever/resolvers/topicResolverSender/diffPeerListCreator.go (hunk 6 above)
Comment (mihaib79) [reply dataRetriever/resolvers/topicResolverSender/diffPeerListCreator.go:77]: Done

File: dataRetriever/errors.go (hunk 5 above)
Comment (mihaib79) [reply dataRetriever/errors.go]: Done

File: p2p/libp2p/netMessenger.go (hunk 3 above)
Comment (mihaib79) [reply p2p/libp2p/netMessenger.go:641]: Done

File: dataRetriever/interface.go (hunk 2 above)
Comment (mihaib79) [reply dataRetriever/interface.go:366]: Done

File: dataRetriever/mock/currentNetworkEpochProviderMock.go (hunk 1 above)
Comment (mihaib79) [reply dataRetriever/mock/currentNetworkEpochProviderMock.go]: Done

File: p2p/libp2p/netMessenger_test.go [hunk 7]
```diff
@@ -866,6 +877,31 @@ func TestLibp2pMessenger_ConnectedPeersOnTopicTwoTopicsShouldWork(t *testing.T)
 	_ = mes4.Close()
 }
 
+//------- ConnectedFullHistoryPeersOnTopic
+
+func TestLibp2pMessenger_ConnectedFullHistoryPeersOnTopicShouldWork(t *testing.T) {
+	_, mes1, mes2 := createMockNetworkOf2()
```
Comment (iulianpascalau) [inline p2p/libp2p/netMessenger_test.go:883]: In order to have a better test for this function I suggest to create a special function called `createMockNetworkOf3` that will connect 3 peers out of which only 2 are full history nodes.

File: dataRetriever/resolvers/topicResolverSender/topicResolverSender.go (hunk 4 above)
Comment (iulianpascalau) [reply dataRetriever/resolvers/topicResolverSender/topicResolverSender.go:125]: Should add unit tests for the new code.

**Review by AdoAdoAdo — DISMISSED** [review]

File: p2p/libp2p/netMessenger.go [hunk 8]
```diff
@@ -637,6 +637,20 @@ func (netMes *networkMessenger) ConnectedPeersOnTopic(topic string) []core.PeerI
...
+	peerList := netMes.ConnectedPeersOnTopic(topic)
+	fullHistoryList := make([]core.PeerID, 0)
+	for _, peer := range peerList {
```
Comment (iulianpascalau) [inline p2p/libp2p/netMessenger.go:644]: Warning here

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: p2p/libp2p/netMessenger.go (hunk 8 above)
Comment (mihaib79) [reply p2p/libp2p/netMessenger.go]: Done

File: p2p/libp2p/netMessenger_test.go (hunk 7 above)
Comment (mihaib79) [reply p2p/libp2p/netMessenger_test.go]: Done

File: p2p/libp2p/netMessenger_test.go [hunk 9]
```diff
@@ -880,26 +910,31 @@ func TestLibp2pMessenger_ConnectedPeersOnTopicTwoTopicsShouldWork(t *testing.T)
...
 
 	_ = mes1.CreateTopic("topic123", false)
 	_ = mes2.CreateTopic("topic123", false)
+	_ = mes3.CreateTopic("topic123", false)
 
 	//wait a bit for topic announcements
 	time.Sleep(time.Second)
 
-	connFullHistoryPeers := mes1.ConnectedFullHistoryPeersOnTopic("topic123")
+	connPeers := mes2.ConnectedPeersOnTopic("topic123")
```
Comment (iulianpascalau) [inline p2p/libp2p/netMessenger_test.go:929]: the asserts should change as follows:
```GO
assert.Equal(t, 2, mes1.ConnectedPeersOnTopic("topic123"))
assert.Equal(t, 1, mes1.ConnectedFullHistoryPeersOnTopic("topic123"))

assert.Equal(t, 2, mes2.ConnectedPeersOnTopic("topic123"))
assert.Equal(t, 1, mes2.ConnectedFullHistoryPeersOnTopic("topic123"))

assert.Equal(t, 2, mes3.ConnectedPeersOnTopic("topic123"))
assert.Equal(t, 1, mes3.ConnectedFullHistoryPeersOnTopic("topic123"))
```

File: p2p/libp2p/netMessenger_test.go [hunk 10]
```diff
@@ -108,6 +108,36 @@ func createMockNetworkOf2() (mocknet.Mocknet, p2p.Messenger, p2p.Messenger) {
...
+
+	_ = netw.LinkAll()
+
+	nscm1 := mock2.NewNetworkShardingCollectorMock()
+	nscm1.UpdatePeerIdSubType(mes1.ID(), core.FullHistoryObserver)
+	nscm1.UpdatePeerIdSubType(mes2.ID(), core.FullHistoryObserver)
+	nscm1.UpdatePeerIdSubType(mes3.ID(), core.RegularPeer)
+	_ = mes1.SetPeerShardResolver(nscm1)
+
+	nscm2 := mock2.NewNetworkShardingCollectorMock()
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply p2p/libp2p/netMessenger_test.go]: Done

File: p2p/libp2p/netMessenger_test.go [hunk 11]
```diff
@@ -880,26 +910,31 @@ func TestLibp2pMessenger_ConnectedPeersOnTopicTwoTopicsShouldWork(t *testing.T)
...
 func TestLibp2pMessenger_ConnectedFullHistoryPeersOnTopicShouldWork(t *testing.T) {
-	_, mes1, mes2 := createMockNetworkOf2()
+	mes1, mes2, mes3 := createMockNetworkOf3()
 
 	adr2 := mes2.Addresses()[0]
 	fmt.Printf("Connecting to %s...\n", adr2)
 
 	_ = mes1.ConnectToPeer(adr2)
-	//connected peers:  1 ----- 2
+	_ = mes3.ConnectToPeer(adr2)
```
[the remark this answers is not in the document]
Comment (mihaib79) [reply p2p/libp2p/netMessenger_test.go:910]: Done

File: p2p/libp2p/netMessenger_test.go (hunk 9 above)
Comment (mihaib79) [reply p2p/libp2p/netMessenger_test.go]: Done

**Review by AdoAdoAdo — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**